templates/start-dev-env.sh

```diff
@@ -0,0 +1,133 @@
+#!/bin/bash
+
+#
+# This script is to create a dev namespace on Staging environment
+#
+PROJECT_NAME=<% .Name %>
+ENVIRONMENT=stage
+ACCOUNT_ID=<% index .Params `accountId` %>
+REGION=<% index .Params `region` %>
+
+# common functions
+function usage() {
+    echo
+    echo "Usage:"
+    echo "  $0 <IAM user> <project id>"
+    echo "    - IAM user: can be found by running 'aws iam get-group --group-name ${PROJECT_NAME}-developer-${ENVIRONMENT} | jq -r .Users[].UserName'"
+    echo "    - project id: can be 001, 002, or whatever id without space"
+    exit 1
+}
+
+function command_exist() {
+    command -v ${1} >& /dev/null
+}
+
+function error_exit() {
+    echo "ERROR : $1"
+    exit 2
+}
+
+function can_i() {
+    command=$1
+    kubectl auth can-i $1 >& /dev/null || error_exit "No permission to $1"
+}
+
+# Start
+USER_NAME=$1
+DEV_PROJECT_ID=$2
+[[ -z "$USER_NAME" ]] && usage
+[[ -z "$DEV_PROJECT_ID" ]] && DEV_PROJECT_ID="001"
+
+echo '[Dev Environment]'
+
+# Validate cluster
+CLUSTER=$(kubectl config current-context)
+[[ ${CLUSTER} == "${PROJECT_NAME}-${ENVIRONMENT}-${REGION}" ]] || error_exit "Your kubernetes context ${CLUSTER} is not proper to run this script"
+echo "  Cluster: ${CLUSTER}"
+
+# Validate secret
+NAMESPACE=${PROJECT_NAME}
+SECRET_NAME=${PROJECT_NAME}
+DEV_SECRET_NAME=dev-${USER_NAME}
+DEV_SECRET_JSON=$(kubectl get secret ${DEV_SECRET_NAME} -n ${NAMESPACE} -o json)
+[[ -z "${DEV_SECRET_JSON}" ]] && error_exit "The secret ${DEV_SECRET_NAME} is not existing. Please check secrets with 'kubectl get secrets'"
+
+# Check installations
+if ! command_exist skaffold || ! command_exist telepresence; then
+    if ! command_exist skaffold; then
+        error_exit "command 'skaffold' not found: please visit https://skaffold.dev/docs/install/"
+    fi
+    if ! command_exist kubectl; then
+        error_exit "command 'telepresence' not found. You can download it at https://www.telepresence.io/reference/install."
+    fi
+fi
+
+# Setup dev namepsace
+DEV_USER_PROJECT=${USER_NAME}-${DEV_PROJECT_ID}
+DEV_NAMESPACE=${DEV_USER_PROJECT}
+kubectl get namespace ${DEV_NAMESPACE} >& /dev/null || \
+    (can_i "create namespace" && kubectl create namespace ${DEV_NAMESPACE} && \
```

Comment:
Can you change this so that it checks all the permissions, reports on any that the user can't do, then exit? Otherwise it will exit on the first one the user can't do.

Reply:
good point, better check all in one time